src/pyhf/constraints.py

```diff
@@ -125,62 +151,82 @@ def _precompute(self):
             try:
                 poisson_constraint_rate_factors.append(parset.factors)
             except AttributeError:
-                poisson_constraint_rate_factors.append(
-                    default_backend.shape(self.par_indices[parslice])
-                )
+                # this seems to be dead code
+                # TODO: add coverage (issue #540)
+                poisson_constraint_rate_factors.append([1.0] * len(thisauxdata))
 
-        if poisson_constraint_rate_indices:
-            poisson_rate_idc = default_backend.concatenate(
-                list(
-                    map(
-                        lambda x: default_backend.astensor(x, dtype='int'),
-                        poisson_constraint_rate_indices,
-                    )
-                )
-            )
-            poisson_rate_fac = default_backend.concatenate(
-                list(
-                    map(
-                        lambda x: default_backend.astensor(x, dtype='float'),
-                        poisson_constraint_rate_factors,
-                    )
-                )
-            )
-            poisson_data = default_backend.concatenate(
-                list(
-                    map(
-                        lambda x: default_backend.astensor(x, dtype='int'),
-                        poisson_constraint_data,
-                    )
-                )
+        if self.parameter_helper.index_selection:
+            self._poisson_data = default_backend.astensor(
+                default_backend.concatenate(poisson_constraint_data), dtype='int'
             )
 
-            self.poisson_data = tensorlib.astensor(
-                default_backend.tolist(poisson_data), dtype='int'
+            _poisson_rate_fac = default_backend.astensor(
+                default_backend.concatenate(poisson_constraint_rate_factors),
+                dtype='float',
             )
-            self.poisson_rate_idc = tensorlib.astensor(
-                default_backend.tolist(poisson_rate_idc), dtype='int'
-            )
-            self.poisson_rate_fac = tensorlib.astensor(
-                default_backend.tolist(poisson_rate_fac), dtype='float'
+            factors = default_backend.reshape(
+                _poisson_rate_fac, (1, -1)
+            )  # (1, normals)
+            self._batched_factors = default_backend.tile(
+                factors, (self.batch_size or 1, 1)
```

Comment:
Why is there a `1` in `(batch_size, 1)` and not `-1` or otherwise?

Reply:
for this data member the unbatched computation reduces to a computation with batch_size = 1

Reply:
Not following. Is there a clear example of what you expect this calculation to do?